modules/standard/IO.chpl

```diff
@@ -3977,125 +3992,209 @@ proc channel.readline(ref arg: ?t): bool throws where t==string || t==bytes {
-      if(maxSize != -1) {
-        this._mark();
-        var lineSize = 0;
-        var got : int;
-        while got!=newLineChar {
-          got = qio_channel_read_byte(false, this._channel_internal);
-          if(got < 0) then break;
-          lineSize+=1;
-          if(lineSize > maxSize){
-                    // The line is longer than was specified so we throw an error
-          this._revert();
-          try this._ch_ioerror(EFORMAT:syserr, "line longer than maxSize in channel.readLine(ref b: bytes)");
-          }
-        }
+  on this.home {
+    try this.lock(); defer { this.unlock(); }
+    param newLineChar = 0x0A; // ascii newline.
+    var maxBytes = if maxSize < 0 then max(int) else maxSize;
+    var nBytes: int = 0;
+    // use the channel's buffering to compute how many bytes we are reading
+    this._mark();
+    var got : int;
+    var err: syserr = ENOERR;
+    var foundNewline = false;
+    while !foundNewline {
+      // read a sigle byte
+      got = qio_channel_read_byte(false, this._channel_internal);
+      if got == -EEOF {
+        // encountered EOF
+        break;
+      } else if got < 0 {
+        // encountered an error so throw
         this._revert();
+        err = -got;
+        try this._ch_ioerror(err, "in channel.readLine(ref b: bytes)");
+        break;
       }
-      var saveStyle: iostyleInternal = this._styleInternal();
-      defer {
-        this._set_styleInternal(saveStyle);
+      nBytes += 1;
+      if got == newLineChar {
+        foundNewline = true;
       }
-      var myStyle = saveStyle.text();
-      myStyle.string_format = QIO_STRING_FORMAT_TOEND;
-      myStyle.string_end = newLineChar;
-      if(maxSize >= 0){
-        try myStyle.max_width_bytes = maxSize.safeCast(uint(32)); // Probably should change maxSize's type to be safe to cast
+      if nBytes > maxBytes {
+        if foundNewline && stripNewline {
+          // don't worry about it since we wouldn't return the newline
+        } else {
+          // The line is longer than was specified so we throw an error
+          this._revert();
+          try this._ch_ioerror(EFORMAT:syserr,
+                   "line longer than maxSize in channel.readLine(ref b: bytes)");
+        }
       }
-      this._set_styleInternal(myStyle);
-      try _readOne(iokind.dynamic, b, origLocale);
     }
-  } catch err: SystemError {
-    if err.err != EEOF then throw err;
-    return false;
-  }
-  // TODO find a more memory efficient way to do this
-  // Maybe have a QIO_STRING_FORMAT_TOEND_DROPEND which drops the last char
-  if(stripNewline){
-    b = b.strip(chars=b"\n", leading = false);
+    this._revert();
+    // now, nBytes includes the newline
+    if foundNewline && stripNewline {
+      // but we don't want to read the newline if stripNewline=true.
+      nBytes -= 1;
+    }
+
+    // now read the data into the bytes
+    err = readStringBytesData(b, this._channel_internal, nBytes, nCodepoints=0);
```

Comment:
nit: can we either:

- make nCodepoints default to something negative in the function's declaration and not pass anything here,
- or, just at least pass something negative here.

Reply:
Sure I will pass -1 here